### PR TITLE
Bump checkerframework to 2.2.2 and stop building from source

### PR DESCRIPTION
**Pull request by davido**
Fixes: #4709.

This change stops building checker framework from the source and consumes
the artifacts from Central. As the consequence, the sources are removed
from the Bazel tree and we discontinue packaging the sources in Bazel
source artifact.

GPL v2 with Classpath exception sources have to be still included in
`third_party:gpl-srcs` rule. The source jars are included in this rule,
that is used in embeded_tools archive:

```
$ unzip -t bazel-genfiles/src/embedded_tools.zip|grep checker_framework
testing: third_party/checker_framework_dataflow/dataflow-2.2.2-sources.jar
testing: third_party/checker_framework_javacutil/javacutil-2.2.2-sources.jar
```

### CODE REVIEW

**Comment by philwo** [comment]
@davido Maybe I'm missing something, but where does the `dataflow-2.2.2.jar` come from? Doesn't seem to be part of this PR?

**Comment by davido** [comment]
@philwo Oh, my bad, sorry, probably I missed to add it to the commit? I will re-add it later today and push a new version of the PR. (The source is just Maven Central, the location that was referenced by @cushon).

Also I'm not sure you really want to merge another 2, Java 9 related commits, They are more or less a RFC.

File: src/main/java/com/google/devtools/build/lib/skyframe/serialization/autocodec/BUILD [hunk 1]
```diff
@@ -11,6 +11,10 @@ filegroup(
 java_library(
     name = "autocodec",
     exported_plugins = [":autocodec-plugin"],
+    javacopts = select({
```
Comment (philwo) [inline src/main/java/com/google/devtools/build/lib/skyframe/serialization/autocodec/BUILD]: @janakdr What's your thought on building autocodec with `--add-modules=jdk.unsupported` for Java 9? Do you think there's a better way?

File: third_party/BUILD [hunk 2]
```diff
@@ -46,8 +44,6 @@ filegroup(
 filegroup(
```
Comment (cushon) [inline third_party/BUILD]: @philwo @janakdr this change probably isn't OK? We still want to include the sources.

File: src/main/java/com/google/devtools/build/lib/skyframe/serialization/autocodec/BUILD (hunk 1 above)
Comment (cushon) [reply src/main/java/com/google/devtools/build/lib/skyframe/serialization/autocodec/BUILD]: I think the Java 9 changes should be made separately from the "bump checkerframework to 2.2.2 and stop building from source" change. The latter is big enough for one commit, and the details former may need more discussion.

File: third_party/checker_framework_dataflow/BUILD [hunk 3]
```diff
@@ -2,25 +2,12 @@ package(default_visibility = ["//visibility:public"])
...
-)
-
-java_library(
+java_import(
     name = "checker_framework_dataflow",
-    srcs = glob(["java/**"]),
-    javacopts = ["-Xep:MissingCasesInEnumSwitch:OFF"],
+    jars = [
+        "dataflow-2.2.2.jar",
+    ],
```
Comment (cushon) [inline third_party/checker_framework_dataflow/BUILD]: Please add the source jars too.

File: third_party/grpc/compiler/src/java_plugin/cpp/java_generator.cpp [hunk 4]
```diff
@@ -1069,9 +1069,8 @@ static void PrintService(const ServiceDescriptor* service,
   GrpcWriteServiceDocComment(p, service);
   p->Print(
       *vars,
-      "@$Generated$(\n"
-      "    value = \"by gRPC proto compiler$grpc_version$\",\n"
-      "    comments = \"Source: $file_name$\")\n"
+      "// Generated by gRPC proto compiler$grpc_version$\n"
+      "// Source: $file_name$\n"
```
Comment (cushon) [inline third_party/grpc/compiler/src/java_plugin/cpp/java_generator.cpp]: This doesn't belong in this PR, it should be a separate change.

Comment (davido) [reply third_party/grpc/compiler/src/java_plugin/cpp/java_generator.cpp]: Done.

File: third_party/checker_framework_dataflow/BUILD (hunk 3 above)
Comment (davido) [reply third_party/checker_framework_dataflow/BUILD]: Done.

File: src/main/java/com/google/devtools/build/lib/skyframe/serialization/autocodec/BUILD (hunk 1 above)
Comment (davido) [reply src/main/java/com/google/devtools/build/lib/skyframe/serialization/autocodec/BUILD]: Done.

File: third_party/BUILD (hunk 2 above)
Comment (davido) [reply third_party/BUILD]: @cushon can you elaborate? Is that because of LGPL license?
The sources must be in tree? Can we merge from source-JAR downloaded from Central?

Comment (janakdr) [reply third_party/BUILD]: I don't see anything now. Are these changes reverted?

Comment (davido) [reply third_party/BUILD]: Yes, I uploaded a new version of the PR. I restored the GPL source packaging. But now the jars are provided and included in `embedded_tools`:

```
  $ bazel build src:embedded_tools
INFO: Analysed target //src:embedded_tools (1 packages loaded).
INFO: Found 1 target...
Target //src:embedded_tools up-to-date:
  bazel-genfiles/src/embedded_tools.zip
INFO: Elapsed time: 0.279s, Critical Path: 0.00s

  $ unzip -t bazel-genfiles/src/embedded_tools.zip | grep checker_framework
    testing: third_party/checker_framework_dataflow/dataflow-2.2.2-sources.jar   OK
    testing: third_party/checker_framework_javacutil/javacutil-2.2.2-sources.jar   OK

```

File: third_party/checker_framework_dataflow/BUILD [hunk 5]
```diff
@@ -4,23 +4,16 @@ licenses(["restricted"])  # GNU GPL v2 with Classpath exception
 
 filegroup(
     name = "srcs",
-    srcs = glob(["**"]),
+    srcs = ["dataflow-2.2.2-sources.jar"],
```
Comment (cushon) [inline third_party/checker_framework_dataflow/BUILD]: These sources get collected into a tar archive by `//:bazel-srcs`. Currently the tar contains the `.java` sources, this changes causes the tar to contain the source jar. I think we want the preserve the current behaviour.

Comment (davido) [reply third_party/checker_framework_dataflow/BUILD]: This PR is a principle change in handling of checker framework. We stop building from the source and start to consume the pre-built artifacts (downloaded from Maven Central).

If we do not build from the source, and do not have sources in our tree, why should we supply sources in `:bazel-srcs` at all? We don't do this neither for `guava`, nor for `error prone` and nor for custom java compiler.

In fact, before this change, the `//:bazel-srcs` tar already contains some source JARs:

```
  $ davido@wizball:~/projects/bazel (master %=)$ tar -tf bazel-bin/bazel-srcs.tar | grep --regexp\="third_party/.*jar$" | grep source
./third_party/asm/asm-6.0-sources.jar
./third_party/asm/asm-analysis-6.0-sources.jar
./third_party/asm/asm-commons-6.0-sources.jar
./third_party/asm/asm-tree-6.0-sources.jar
./third_party/asm/asm-util-6.0-sources.jar
./third_party/java/jarjar/jarjar-core/src/test/resources/enumtest.jar
```

After this change we have two more:

```
[...]
./third_party/checker_framework_dataflow/dataflow-2.2.2-sources.jar
./third_party/checker_framework_javacutil/javacutil-2.2.2-sources.jar
```

So, if you would argue, that we should stop include source JARs for checker framework in `//:bazel-srcs` artifact I think would agree with this point.

A different point is third-party:gpl-srcs. There we also include source JARs for the checker-framework that is part of embeded zip. But there we also have JARs, Moreover, I also think, that a follow-up change could be to stop building `jFormatString` from the source (another LGPL licensed library) and consume pre-built binary from the Central:

```
<dependency>
    <groupId>com.google.code.findbugs</groupId>
    <artifactId>jFormatString</artifactId>
    <version>3.0.0</version>
</dependency>

```

Comment (davido) [reply third_party/checker_framework_dataflow/BUILD]: > Should we rather remove checker framework source JARs from `//:bazel-srcs` entirely?

I discontinued packaging checker framework in `//:bazel-srcs` in the updated PR, and extended the commit message:

```
This change stops building checker framework from the source and consumes
the artifacts from Central. As the consequence, the sources are removed
from the Bazel tree and we discontinue packaging the sources in Bazel
source artifact.

LGPL sources are still have to be included in embeded_tools archive. The
source jars are included in embedded_tools archive now:

$ unzip -t bazel-genfiles/src/embedded_tools.zip|grep checker_framework
testing: third_party/checker_framework_dataflow/dataflow-2.2.2-sources.jar
testing: third_party/checker_framework_javacutil/javacutil-2.2.2-sources.jar
```

Comment (davido) [reply third_party/checker_framework_dataflow/BUILD]: > Moreover, I also think, that a follow-up change could be to stop building jFormatString from the source (another LGPL licensed library) and consume pre-built binary from the Central ...

Done in https://github.com/bazelbuild/bazel/pull/4843

**Comment by lfpino** [comment]
Hi @davido, Bazel sheriff here, what's the status of this pull request? It hasn't been updated for more than a week so it'd be good to understand if it's still ongoing or safe to close. Thanks.
